tests/test_cmake_build/test.py

```diff
@@ -3,5 +3,7 @@
 
 import test_cmake_build
 
+__file__  # Test this is properly set
```

Comment:
What do you think about making this slightly stronger (and look more like a test)? E.g.
```
assert isinstance(__file__, str)  # Test this is properly set.
```

Reply:
It's actually confusing if it's suppose to a byte or str or if that itself is implementation dependant. Were you able to find the spec that says it's suppose to a str? And a unicode str or an ascii str at that?

Reply:
I just left it as is since if it's a bug, it's a bug with our entire function signature for eval.

Reply:
I knew it! 
```
isinstance(__file__, str) == False on Python2
```

Reply:
Oh, because it is unicode even on Python 2? — I think that counts as a bug, too!
Although not a surprising one given how pybind11 works internally. I think in that case we want to keep the test as is (isinstance str), but encode in the C++ code.
Give me a moment to look around a bit what the expections are for __file__ with Python 2.

Reply:
A bit random, but this looks very similar:

https://github.com/pypa/pip/pull/1538/files

Reply:
Maybe more pragmatically: guard the new C++ code with `#ifdef` for Python 3, and the test with `not env.PY2`.
I.e. we're simply leaving Python 2 behind (instead of going through hoops for something that is practically dead already).

Reply:
I don't like that because it should be defined in Python 2, the previous code was checking for NameError.

Reply:
> 
> 
> A bit random, but this looks very similar:
> 
> https://github.com/pypa/pip/pull/1538/files

Nah, this is same code snippit is suppose to be done in the Python2 side to convert __file__ to the appropriate format.

Reply:
I'm good with either:
* status quo for Python 2 (carry forward existing shortcoming),
* or fixing properly.
For the latter, is there a Python C API for `sys.getfilesystemencoding()`? If not we have to import, getattr, call.